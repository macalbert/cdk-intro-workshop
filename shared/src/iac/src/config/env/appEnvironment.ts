export enum AppEnvironment {
    Production = "Production",
    Development = "Development",
    Staging = "Staging",
    GithubAction = "GithubAction",
    Test = "Test",
}
