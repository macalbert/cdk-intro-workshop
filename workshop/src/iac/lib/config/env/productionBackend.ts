import {
    ApiGatewayStack,
    type ApiGatewayStackProps,
    AppEnvironment,
    ApiLambdaStack,
    type ApiLambdaStackProps,
    S3Stack,
    type S3StackProps,
    StackBuildPart,
    type StackBuildProps,
} from "@m47/shared-iac";
import { type Stack } from "aws-cdk-lib";
import { SubnetType } from "aws-cdk-lib/aws-ec2";
import type { IFunction } from "aws-cdk-lib/aws-lambda";

export interface ProductionBackendProps extends StackBuildProps {
    clusterName: string;
    serviceName: string;
    exposedPort: number;
    sourceCodePath: string;
    dockerfileApi: string;
}

export class ProductionBackend extends StackBuildPart {
    props: ProductionBackendProps;

    environment = AppEnvironment.Production;

    constructor(buildProps: ProductionBackendProps) {
        super(buildProps);

        this.props = buildProps;
        this.props.githubRepo = this.formatRepoNameForCloudFormation();
    }

    build(): Stack[] {
        
        const apiLambdaStack = this.createApiLambdaStack(
            "ApiLambda",
            this.props.dockerfileApi,
            "api",
        );

        const apiGatewayStack = this.createApiGateway(
            apiLambdaStack.lambdaFunction,
            "ApiGateway",
            "api-cdk-workshop",
        );

        return [
            apiLambdaStack,
            apiGatewayStack,
        ];
    }

    private createApiLambdaStack(
        name: string,
        dockerfile: string,
        subdomain: string,
    ): ApiLambdaStack {
        const lambdaExportProps: ApiLambdaStackProps = {
            name: name,
            githubRepo: this.props.githubRepo,
            pathDockerFile: this.props.sourceCodePath,
            env: this.props.env,
            vpc: this.props.vpc,
            envName: AppEnvironment.Production,
            timeoutSeconds: 29,
            memorySizeMbs: 1024,
            vpcSubnets: {
                subnetType: SubnetType.PRIVATE_WITH_EGRESS,
            },
            stackName: `${this.props.githubRepo}-${name}`,
            functionName: `${this.props.githubRepo}-${subdomain}`.toLowerCase(),
            dockerFile: dockerfile,
        };

        const stack = new ApiLambdaStack(this.props.scope, lambdaExportProps);

        return stack;
    }

    private createApiGateway(
        lambdaFunction: IFunction,
        name: string,
        subdomain: string,
    ): ApiGatewayStack {
        const apiProps: ApiGatewayStackProps = {
            env: this.props.env,
            name: name,
            envName: AppEnvironment.Production,
            lambdaFunction: lambdaFunction,
            stackName: `${this.props.githubRepo}-${name}`,
            subdomain: subdomain,
            githubRepo: this.props.githubRepo,
            certificateArn: `arn:aws:acm:us-east-1:${this.props.env.account}:certificate/be63062d-5316-47af-9f94-819c1dc02853`,
            domain: "m47.io",
        };

        return new ApiGatewayStack(this.props.scope, apiProps);
    }
}
