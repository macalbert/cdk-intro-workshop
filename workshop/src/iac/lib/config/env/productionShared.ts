import {
    AppEnvironment,
    PipelineStack,
    type PipelineStackProps,
    RdsStack,
    type RdsStackProps,
    StackBuildPart,
    type StackBuildProps,
    WindowsBastionStack,
    type WindowsBastionStackProps,
} from "@m47/shared-iac";
import { RemovalPolicy, type Stack } from "aws-cdk-lib";
import { InstanceType, InstanceClass, InstanceSize } from "aws-cdk-lib/aws-ec2";
import { PostgresEngineVersion } from "aws-cdk-lib/aws-rds";

export class ProductionShared extends StackBuildPart {
    props: StackBuildProps;

    constructor(buildProps: StackBuildProps) {
        super(buildProps);

        this.props = buildProps;
    }

    build(): Stack[] {
        const pipelineStack = this.createPipelineStack();
        const rdsStack = this.createRdsStack();
        const createBastionStack = this.createBastionStack();

        return [
            rdsStack,
            pipelineStack,
            createBastionStack,
        ];
    }

    private createPipelineStack(): PipelineStack {
        const pipeline: PipelineStackProps = {
            name: "Pipeline",
            branch: this.props.branch,
            githubRepo: this.props.githubRepo,
            envName: AppEnvironment.Production,
            secretTokenArn: `arn:aws:secretsmanager:${this.props.env.region}:${this.props.env.account}:secret:github-access-token-secret-dth2kY`,
            testBuildSpec: [
                "workshop/src/iac/buildspecs/test/iac.yml",
                "workshop/src/iac/buildspecs/test/backend.yml",
            ],
            deployBuildSpec: [
                "workshop/src/iac/buildspecs/production/backendApi.yml",
            ],
            testProjectName: `${this.props.githubRepo}-Tests`,
            vpc: this.props.vpc,
            env: this.props.env,
            bucketRemovalPolicy: RemovalPolicy.DESTROY,
            domain: "workshop.com",
            stackName: `${this.formatRepoNameForCloudFormation()}-CodePipeline`,
            manualApproval: false,
            shouldCreateNewSlackChannel: false,
        };

        return new PipelineStack(this.props.scope, pipeline);
    }

    private createRdsStack(): RdsStack {
        const props: RdsStackProps = {
            env: this.props.env,
            vpc: this.props.vpc,
            databaseName: "workshop",
            databaseUserName: "masteruser",
            backupRetentionDays: 14,
            storageSizeGb: 100,
            monitoringIntervalSeconds: 60,
            engineVersion: PostgresEngineVersion.VER_17_4,
            instanceType: InstanceType.of(InstanceClass.T3, InstanceSize.MICRO),
            githubRepo: this.props.githubRepo,
            envName: AppEnvironment.Production,
            stackName: `${this.formatRepoNameForCloudFormation()}-rds`,
            name: "RdsPostgres",
        };

        return new RdsStack(this.props.scope, props);
    }

    private createBastionStack(): WindowsBastionStack {
        const windowsBastionStackProps: WindowsBastionStackProps = {
            vpc: this.props.vpc,
            githubRepo: this.props.githubRepo,
            envName: AppEnvironment.Production,
            name: "bastion",
            stackName: `${this.formatRepoNameForCloudFormation()}-Bastion`,
            env: this.props.env,
        };

        return new WindowsBastionStack(
            this.props.scope,
            windowsBastionStackProps,
        );
    }
}
