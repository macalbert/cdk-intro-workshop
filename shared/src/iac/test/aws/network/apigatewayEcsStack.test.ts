import { App, Stack } from "aws-cdk-lib";
import { Template } from "aws-cdk-lib/assertions";
import {
	ApiGatewayEcsStack,
	type ApiGatewayEcsStackProps,
} from "../../../src/aws/network/apigatewayEcsStack";
import { AppEnvironment } from "../../../src/config/env/appEnvironment";

describe("ApigatewayStack", () => {
	const env = {
		account: "account",
		region: "eu-west-1",
	};

	test("Should_CreateNewApiGateway_When_StackIsCalled", () => {
		// Arrange
		const stack = new Stack(new App(), "apigatewayEcsStacktest", {
			env: env,
		});
		const props = createApigatewayEcsStackProps();

		// Act
		const actual = new ApiGatewayEcsStack(stack, props);

		// Assert
		const template = Template.fromStack(actual).toJSON();

		for (const key in template.Resources) {
			const resource = template.Resources[key];

			if (
				typeof key === "string" &&
				key.startsWith("LambdaRestApitestapigatewayDeployment")
			) {
				delete template.Resources[key];
			}
			if (resource.Properties.DeploymentId?.Ref) {
				delete resource.Properties.DeploymentId.Ref;
			}
		}

		expect(template).toMatchSnapshot("apigatewayEcsStacktest");
	});

	function createApigatewayEcsStackProps(): ApiGatewayEcsStackProps {
		return {
			env: env,
			subdomain: "shared-api-test",
			name: "GatewayApiStack",
			envName: AppEnvironment.Test,
			githubRepo: "test",
			stackName: `test-apigateway`,
			loadBalancerArn: `arn:aws:elasticloadbalancing:${env.region}:${env.account}:loadbalancer/somePath`,
			domain: "some.test",
			certificateArn: `arn:aws:acm:us-east-1:${env.account}:certificate/someId`,
		};
	}
});
