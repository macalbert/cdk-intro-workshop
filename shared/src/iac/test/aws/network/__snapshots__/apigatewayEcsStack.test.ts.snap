// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`ApigatewayStack Should_CreateNewApiGateway_When_StackIsCalled: apigatewayEcsStacktest 1`] = `
{
  "Outputs": {
    "EcsRestApitestapigatewayEndpoint6DC0813A": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "EcsRestApitestapigatewayB8EC6526",
            },
            ".execute-api.eu-west-1.",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "EcsRestApitestapigatewayDeploymentStageprodBC1873B4",
            },
            "/",
          ],
        ],
      },
    },
    "apiUrl": {
      "Value": {
        "Fn::Join": [
          "",
          [
            "https://",
            {
              "Ref": "EcsRestApitestapigatewayB8EC6526",
            },
            ".execute-api.eu-west-1.",
            {
              "Ref": "AWS::URLSuffix",
            },
            "/",
            {
              "Ref": "EcsRestApitestapigatewayDeploymentStageprodBC1873B4",
            },
            "/",
          ],
        ],
      },
    },
    "recordUrl": {
      "Value": "https://shared-api-test.some.test",
    },
  },
  "Parameters": {
    "BootstrapVersion": {
      "Default": "/cdk-bootstrap/hnb659fds/version",
      "Description": "Version of the CDK Bootstrap resources in this environment, automatically retrieved from SSM Parameter Store. [cdk:skip]",
      "Type": "AWS::SSM::Parameter::Value<String>",
    },
  },
  "Resources": {
    "ARecordtestapigatewayD480FC99": {
      "Properties": {
        "AliasTarget": {
          "DNSName": {
            "Fn::GetAtt": [
              "EcsRestApitestapigatewayCustomDomainD5141EBD",
              "DistributionDomainName",
            ],
          },
          "HostedZoneId": {
            "Fn::GetAtt": [
              "EcsRestApitestapigatewayCustomDomainD5141EBD",
              "DistributionHostedZoneId",
            ],
          },
        },
        "HostedZoneId": "DUMMY",
        "Name": "shared-api-test.some.test.",
        "Type": "A",
      },
      "Type": "AWS::Route53::RecordSet",
    },
    "ApiGatewayAccessLogstestapigateway87AB4249": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "RetentionInDays": 365,
      },
      "Type": "AWS::Logs::LogGroup",
      "UpdateReplacePolicy": "Delete",
    },
    "ApiGatewayErrorAlarmtestapigateway59DB36BD": {
      "Properties": {
        "ActionsEnabled": true,
        "AlarmActions": [
          {
            "Ref": "ApiGatewayErrorTopictestapigatewayBD34613D",
          },
        ],
        "AlarmDescription": "API Gateway Error Alarm",
        "ComparisonOperator": "GreaterThanOrEqualToThreshold",
        "Dimensions": [
          {
            "Name": "ApiName",
            "Value": "EcsRestApi-test-apigateway",
          },
          {
            "Name": "Stage",
            "Value": {
              "Ref": "EcsRestApitestapigatewayDeploymentStageprodBC1873B4",
            },
          },
        ],
        "EvaluationPeriods": 1,
        "MetricName": "5XXError",
        "Namespace": "AWS/ApiGateway",
        "Period": 300,
        "Statistic": "Sum",
        "Threshold": 1,
      },
      "Type": "AWS::CloudWatch::Alarm",
    },
    "ApiGatewayErrorTopictestapigatewayBD34613D": {
      "Properties": {
        "DisplayName": "API Gateway Error Topic",
      },
      "Type": "AWS::SNS::Topic",
    },
    "EcsRestApitestapigatewayAccountC6E06284": {
      "DeletionPolicy": "Retain",
      "DependsOn": [
        "EcsRestApitestapigatewayB8EC6526",
      ],
      "Properties": {
        "CloudWatchRoleArn": {
          "Fn::GetAtt": [
            "EcsRestApitestapigatewayCloudWatchRole4B4E7886",
            "Arn",
          ],
        },
      },
      "Type": "AWS::ApiGateway::Account",
      "UpdateReplacePolicy": "Retain",
    },
    "EcsRestApitestapigatewayB8EC6526": {
      "Properties": {
        "BinaryMediaTypes": [
          "multipart/form-data",
        ],
        "Description": "API Gateway ECS Fargate",
        "Name": "EcsRestApi-test-apigateway",
        "Tags": [
          {
            "Key": "Environment",
            "Value": "Test",
          },
          {
            "Key": "Project",
            "Value": "test",
          },
          {
            "Key": "StackId",
            "Value": "m47-test-gatewayapistack-test-stack",
          },
        ],
      },
      "Type": "AWS::ApiGateway::RestApi",
    },
    "EcsRestApitestapigatewayCloudWatchRole4B4E7886": {
      "DeletionPolicy": "Retain",
      "Properties": {
        "AssumeRolePolicyDocument": {
          "Statement": [
            {
              "Action": "sts:AssumeRole",
              "Effect": "Allow",
              "Principal": {
                "Service": "apigateway.amazonaws.com",
              },
            },
          ],
          "Version": "2012-10-17",
        },
        "ManagedPolicyArns": [
          {
            "Fn::Join": [
              "",
              [
                "arn:",
                {
                  "Ref": "AWS::Partition",
                },
                ":iam::aws:policy/service-role/AmazonAPIGatewayPushToCloudWatchLogs",
              ],
            ],
          },
        ],
        "Tags": [
          {
            "Key": "Environment",
            "Value": "Test",
          },
          {
            "Key": "Project",
            "Value": "test",
          },
          {
            "Key": "StackId",
            "Value": "m47-test-gatewayapistack-test-stack",
          },
        ],
      },
      "Type": "AWS::IAM::Role",
      "UpdateReplacePolicy": "Retain",
    },
    "EcsRestApitestapigatewayCustomDomainD5141EBD": {
      "Properties": {
        "CertificateArn": "arn:aws:acm:us-east-1:account:certificate/someId",
        "DomainName": "shared-api-test.some.test",
        "EndpointConfiguration": {
          "Types": [
            "EDGE",
          ],
        },
        "Tags": [
          {
            "Key": "Environment",
            "Value": "Test",
          },
          {
            "Key": "Project",
            "Value": "test",
          },
          {
            "Key": "StackId",
            "Value": "m47-test-gatewayapistack-test-stack",
          },
        ],
      },
      "Type": "AWS::ApiGateway::DomainName",
    },
    "EcsRestApitestapigatewayCustomDomainMapapigatewayEcsStacktestm47testgatewayapistackteststackEcsRestApitestapigateway5D47165EAC4C9551": {
      "Properties": {
        "DomainName": {
          "Ref": "EcsRestApitestapigatewayCustomDomainD5141EBD",
        },
        "RestApiId": {
          "Ref": "EcsRestApitestapigatewayB8EC6526",
        },
        "Stage": {
          "Ref": "EcsRestApitestapigatewayDeploymentStageprodBC1873B4",
        },
      },
      "Type": "AWS::ApiGateway::BasePathMapping",
    },
    "EcsRestApitestapigatewayDeployment62CBAB02a43d96a34d9c81891b28160d49fd5bc2": {
      "DependsOn": [
        "EcsRestApitestapigatewayproxyANY5EF5721D",
        "EcsRestApitestapigatewayproxyDC3B8397",
      ],
      "Metadata": {
        "aws:cdk:do-not-refactor": true,
      },
      "Properties": {
        "Description": "API Gateway ECS Fargate",
        "RestApiId": {
          "Ref": "EcsRestApitestapigatewayB8EC6526",
        },
      },
      "Type": "AWS::ApiGateway::Deployment",
    },
    "EcsRestApitestapigatewayDeploymentStageprodBC1873B4": {
      "DependsOn": [
        "EcsRestApitestapigatewayAccountC6E06284",
      ],
      "Properties": {
        "AccessLogSetting": {
          "DestinationArn": {
            "Fn::GetAtt": [
              "ApiGatewayAccessLogstestapigateway87AB4249",
              "Arn",
            ],
          },
          "Format": "$context.identity.sourceIp $context.identity.caller $context.identity.user [$context.requestTime] "$context.httpMethod $context.resourcePath $context.protocol" $context.status $context.responseLength $context.requestId",
        },
        "DeploymentId": {},
        "RestApiId": {
          "Ref": "EcsRestApitestapigatewayB8EC6526",
        },
        "StageName": "prod",
        "Tags": [
          {
            "Key": "Environment",
            "Value": "Test",
          },
          {
            "Key": "Project",
            "Value": "test",
          },
          {
            "Key": "StackId",
            "Value": "m47-test-gatewayapistack-test-stack",
          },
        ],
      },
      "Type": "AWS::ApiGateway::Stage",
    },
    "EcsRestApitestapigatewayproxyANY5EF5721D": {
      "Properties": {
        "AuthorizationType": "NONE",
        "HttpMethod": "ANY",
        "Integration": {
          "ConnectionId": {
            "Ref": "VpcLinktestapigatewayCD7D21E9",
          },
          "ConnectionType": "VPC_LINK",
          "IntegrationHttpMethod": "ANY",
          "RequestParameters": {
            "integration.request.path.proxy": "method.request.path.proxy",
          },
          "Type": "HTTP_PROXY",
          "Uri": "http://my-load-balancer-1234567890.us-west-2.elb.amazonaws.com/{proxy}",
        },
        "RequestParameters": {
          "method.request.path.proxy": true,
        },
        "ResourceId": {
          "Ref": "EcsRestApitestapigatewayproxyDC3B8397",
        },
        "RestApiId": {
          "Ref": "EcsRestApitestapigatewayB8EC6526",
        },
      },
      "Type": "AWS::ApiGateway::Method",
    },
    "EcsRestApitestapigatewayproxyDC3B8397": {
      "Properties": {
        "ParentId": {
          "Fn::GetAtt": [
            "EcsRestApitestapigatewayB8EC6526",
            "RootResourceId",
          ],
        },
        "PathPart": "{proxy+}",
        "RestApiId": {
          "Ref": "EcsRestApitestapigatewayB8EC6526",
        },
      },
      "Type": "AWS::ApiGateway::Resource",
    },
    "VpcLinktestapigatewayCD7D21E9": {
      "DeletionPolicy": "Delete",
      "Properties": {
        "Description": "VPC Link to the NLB",
        "Name": "VpcLink-test-apigateway",
        "TargetArns": [
          "arn:aws:elasticloadbalancing:us-west-2:123456789012:loadbalancer/network/my-load-balancer/50dc6c495c0c9188",
        ],
      },
      "Type": "AWS::ApiGateway::VpcLink",
      "UpdateReplacePolicy": "Delete",
    },
  },
  "Rules": {
    "CheckBootstrapVersion": {
      "Assertions": [
        {
          "Assert": {
            "Fn::Not": [
              {
                "Fn::Contains": [
                  [
                    "1",
                    "2",
                    "3",
                    "4",
                    "5",
                  ],
                  {
                    "Ref": "BootstrapVersion",
                  },
                ],
              },
            ],
          },
          "AssertDescription": "CDK bootstrap stack version 6 required. Please run 'cdk bootstrap' with a recent version of the CDK CLI.",
        },
      ],
    },
  },
}
`;
